repo/blob/azure/azure_pit.go

```diff
@@ -0,0 +1,199 @@
+				return nil
+			}
+
+			if versionTime.After(versionLimit) {
+				return vs[:i]
+			}
+		}
+	}
+
+	return vs
+}
+
+// listBlobVersions returns a list of blob versions but the blob is deleted, it returns Azure's delete marker version but excludes
+// the Kopia delete marker version that is used to get around immutability protections.
+func (az *azPointInTimeStorage) listBlobVersions(ctx context.Context, prefix blob.ID, callback func(vm versionMetadata) error) error {
+	prefixStr := az.getObjectNameString(prefix)
+
+	pager := az.service.NewListBlobsFlatPager(az.container, &azblob.ListBlobsFlatOptions{
+		Prefix: &prefixStr,
+		Include: azblob.ListBlobsInclude{
+			Metadata:            true,
+			DeletedWithVersions: true, // this shows DeleteMarkers aka blobs with HasVersionsOnly set to true
+			Versions:            true,
+		},
+	})
+
+	for pager.More() {
+		page, err := pager.NextPage(ctx)
+		if err != nil {
+			return translateError(err)
+		}
+
+		for _, it := range page.Segment.BlobItems {
+			vm := az.getVersionedBlobMeta(it)
+
+			if err := callback(vm); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (az *azPointInTimeStorage) getVersionedMetadata(ctx context.Context, blobID blob.ID) (versionMetadata, error) {
+	var vml []versionMetadata
+
+	if err := az.getBlobVersions(ctx, blobID, func(vm versionMetadata) error {
+		if !vm.Timestamp.After(az.pointInTime) {
+			vml = append(vml, vm)
+		}
+
+		return nil
+	}); err != nil {
+		return versionMetadata{}, errors.Wrapf(err, "could not get version metadata for blob %s", blobID)
+	}
+
+	if v, found := newestAtUnlessDeleted(vml, az.pointInTime); found {
+		return v, nil
+	}
+
+	return versionMetadata{}, blob.ErrBlobNotFound
+}
+
+// isAzureDeleteMarker checks for Azure created delete markers.
+func (az *azPointInTimeStorage) isAzureDeleteMarker(it *azblobmodels.BlobItem) bool {
+	var isDeleteMarker bool
+	// HasVersionsOnly - Indicates that this root blob has been deleted
+	if it.HasVersionsOnly != nil {
+		isDeleteMarker = *it.HasVersionsOnly
+	}
+
+	return isDeleteMarker
+}
+
+// maybePointInTimeStore wraps s with a point-in-time store when s is versioned
+// and a point-in-time value is specified. Otherwise, s is returned.
+func maybePointInTimeStore(ctx context.Context, s *azStorage, pointInTime *time.Time) (blob.Storage, error) {
+	if pit := s.Options.PointInTime; pit == nil || pit.IsZero() {
+		return s, nil
+	}
+
+	// IsImmutableStorageWithVersioning is needed for PutBlob with RetentionPeriod being set.
+	props, err := s.service.ServiceClient().NewContainerClient(s.container).GetProperties(ctx, nil)
+	if err != nil {
+		return nil, errors.Wrapf(err, "could not get determine if container '%s' supports versioning", s.container)
+	}
+
+	if props.IsImmutableStorageWithVersioningEnabled == nil || !*props.IsImmutableStorageWithVersioningEnabled {
+		return nil, errors.Errorf("cannot create point-in-time view for non-versioned bucket '%s'", s.container)
+	}
+
```

Comment:
Why? Properties is a fast check and a common way to check if a blob or container exists also. I was tempted to replace the `ListBlobs` in `New` with `GetProperties` for the same reason

Reply:
Because (it seems that) it'd be good to reduce the set of permissions needed by kopia, in particular avoid requiring permissions to check bucket properties.

What is needed for kopia to work, is for the credentials to have the permissions to perform the actual operations, (e.g., list with versions), and not necessarily the permissions for checking bucket properties.
Anyway, this is something that we can look into later, separate from this PR.